File: Power Query/400-499/408/PQ_408.py
import pandas as pd

path = "400-499/408/PQ_Challenge_408.xlsx"
input = pd.read_excel(path, usecols="A:C", nrows=21)
test = pd.read_excel(path, usecols="E:F", nrows=3)

result = (
    input.join(
        input.pop("Data")
        .str.split(";", expand=True)
        .set_axis(["Quantity", "LocalPrice", "Currency", "Status"], axis=1)
    )
    .astype({"Quantity": float, "LocalPrice": float})
    .assign(
        USD=lambda x: x.LocalPrice * x.Currency.map({"EUR": 1.1, "GBP": 1.3}).fillna(1)
    )
    .query("Status == 'Completed'")
    .sort_values(["Category", "Date"])
    .groupby("Category")
    .tail(3)
    .assign(wx=lambda x: x.Quantity * x.USD)
    .groupby("Category", as_index=False)
    .agg(**{"Weighted Avg USD": ("wx", "sum"), "Quantity": ("Quantity", "sum")})
    .assign(**{"Weighted Avg USD": lambda x: x["Weighted Avg USD"] / x.Quantity})
    .drop(columns="Quantity")
)

# Only one value matches

File: Power Query/400-499/408/PQ_408.R
library(tidyverse)
library(readxl)

path <- "400-499/408/PQ_Challenge_408.xlsx"
input <- read_excel(path, range = "A1:C22")
test <- read_excel(path, range = "E1:F4")

result <- input %>%
  separate_wider_delim(
    Data,
    names = c("Quantity", "LocalPrice", "Currency", "Status"),
    delim = ";"
  ) %>%
  mutate(
    Quantity = as.numeric(Quantity),
    LocalPrice = as.numeric(LocalPrice)
  ) %>%
  mutate(
    USD = case_when(
      Currency == "EUR" ~ LocalPrice * 1.1,
      Currency == "GBP" ~ LocalPrice * 1.3,
      TRUE ~ LocalPrice
    )
  ) %>%
  filter(Status == "Completed") %>%
  arrange(Category, Date) %>%
  slice_tail(n = 3, by = Category) %>%
  summarise(
    `Weighted Avg USD` = sum(Quantity * USD) / sum(Quantity),
    .by = Category
  )

# only one value matches
